File: src-tauri/src/diagnostics.rs
//! Why the application cannot run.
//!
//! Q124 fixes the boot diagnostic set; ticket 26 adds `workspace_relocation_failed`
//! for an interrupted or partial workspace move; ticket 29 adds `port_bind_failed`,
//! `ffmpeg_missing`, and `ffmpeg_unrunnable` for packaged-only startup failures.
//! Q41 explains the shape: real causes plus `unknown`, because an unmapped failure
//! must not render a blank screen. `unknown` carries the raw error in `detail`.
//!
//! # These are not API error codes
//!
//! There is a second enum — `ApiErrorCode` in
//! `backend/src/outreachos_backend/core/errors.py` — with `validation_error`,
//! `not_found`, `conflict`, `workspace_error`, `unauthorized`, and
//! `internal_error`. **The two namespaces share no values and nothing is ever
//! both.** These codes describe why the *application* cannot run; those
//! describe why a *request* failed.
//!
//! Q124 predicts the temptation to reach across arrives around P4. It is
//! called out here, and in the Python enum, so that whoever feels it finds the
//! answer before writing the mapping.
//!
//! Two merges were considered and rejected:
//!
//! - `workspace_locked` into `workspace_unwritable`: different causes, and
//!   more importantly different action sets — take over, versus pick somewhere
//!   else. Merging means the screen either offers "Take over" on a permissions
//!   failure or hides it on a lock.
//! - `sidecar_exited` split by at-boot versus mid-session: same condition,
//!   different surrounding state. The boot machine already knows whether there
//!   is a route to return to, so the screen selects its title from that.
//!   Splitting an enum to carry a fact the state already holds is how enums
//!   start drifting.

use serde::Serialize;

// Defined complete in one place because Q124 settles it as one enum of nine,
// and splitting the definition across the checkpoints that happen to construct
// each variant would make the set impossible to read.
//
// Constructed as: `sidecar_*` and `handshake_timeout` here in checkpoint 3;
// `workspace_*` and `unknown` in checkpoint 4; `migration_failed` and
// `database_newer_than_app` in checkpoint 5. **Delete this attribute in
// checkpoint 5** — once all nine are reachable it is hiding real dead code
// rather than tolerating a half-built enum.
#[allow(dead_code)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum DiagnosticCode {
    /// The stored pointer no longer resolves to a directory. Offers
    /// "Forget this workspace" — Q40, because otherwise this is a permanent
    /// boot loop with no exit that does not involve editing app-data by hand.
    WorkspaceMissing,
    /// The write probe failed. Usually `Program Files` under a non-admin user.
    WorkspaceUnwritable,
    /// A live or foreign-host `.oos-lock`. Offers "Take over".
    WorkspaceLocked,
    /// The process could not be started at all — a missing interpreter in dev,
    /// a missing sidecar executable in a packaged build.
    SidecarSpawnFailed,
    /// It started and then died. At boot or mid-session; same code.
    SidecarExited,
    /// Q76's budgets: the `@@OOS` port line within 10s, or `/health` 200
    /// within a further 20s. One code for both misses, with `detail`
    /// distinguishing them — Q124.
    HandshakeTimeout,
    /// Ticket 29: the sidecar exited before reporting a port because binding
    /// `127.0.0.1:0` failed. Distinct from a generic handshake timeout so the
    /// diagnostics screen can name it.
    PortBindFailed,
    /// Ticket 29: bundled FFmpeg binaries are absent from the path the
    /// packaged build was given.
    FfmpegMissing,
    /// Ticket 29: FFmpeg binaries are present but cannot be executed.
    FfmpegUnrunnable,
    /// Alembic raised, or the pre-migration backup failed. Q85 folds backup
    /// failure in here rather than adding a code, with `detail` naming it and
    /// an explicit "Proceed without backup" action.
    MigrationFailed,
    /// The database is at a revision this build does not know. Detected
    /// *before* upgrading — Q84.
    DatabaseNewerThanApp,
    /// Ticket 26: a workspace move or fresh init failed after the sidecar was
    /// stopped. The pointer still references the original workspace.
    WorkspaceRelocationFailed,
    /// The fallback. Never rendered as a blank screen; `detail` carries the
    /// original error text verbatim.
    Unknown,
}

/// A diagnostic as the frontend receives it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Diagnostic {
    pub code: DiagnosticCode,
    /// One sentence, written for a user, naming what failed.
    pub message: String,
    /// The technical cause: captured stderr, an OS error, a revision id.
    /// Rendered in a monospace block, not a toast.
    pub detail: Option<String>,
}

impl Diagnostic {
    pub fn new(code: DiagnosticCode, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            detail: None,
        }
    }

    pub fn with_detail(mut self, detail: impl Into<String>) -> Self {
        let detail = detail.into();
        if !detail.trim().is_empty() {
            self.detail = Some(detail);
        }
        self
    }
}

/// Map a degraded `/health` payload onto a boot diagnostic.
pub fn diagnose_degraded_health(code: Option<&str>, detail: Option<&str>) -> Diagnostic {
    let detail_text = detail.filter(|value| !value.trim().is_empty());

    match code {
        Some("workspace_locked") => Diagnostic::new(
            DiagnosticCode::WorkspaceLocked,
            "The workspace is in use by another copy of OutreachOS.",
        )
        .with_detail(
            detail_text
                .unwrap_or("Another copy of OutreachOS appears to have this workspace open."),
        ),

        Some("migration_failed") => Diagnostic::new(
            DiagnosticCode::MigrationFailed,
            "The database could not be updated.",
        )
        .with_detail(detail_text.unwrap_or(
            "OutreachOS could not apply a required database change. \
                 Your data has not been modified.",
        )),

        Some("database_newer_than_app") => Diagnostic::new(
            DiagnosticCode::DatabaseNewerThanApp,
            "This workspace was made by a newer version of OutreachOS.",
        )
        .with_detail(detail_text.unwrap_or(
            "The database in this workspace uses a newer format than this copy understands.",
        )),

        Some("ffmpeg_missing") => Diagnostic::new(
            DiagnosticCode::FfmpegMissing,
            "The bundled video tools are missing.",
        )
        .with_detail(detail_text.unwrap_or(
            "OutreachOS could not find FFmpeg or FFprobe where the installer placed them.",
        )),

        Some("ffmpeg_unrunnable") => Diagnostic::new(
            DiagnosticCode::FfmpegUnrunnable,
            "The bundled video tools could not be run.",
        )
        .with_detail(detail_text.unwrap_or(
            "FFmpeg or FFprobe is present but failed to start. The log below may show why.",
        )),

        _ => Diagnostic::new(
            DiagnosticCode::Unknown,
            "The backend started in a degraded state.",
        )
        .with_detail(detail_text.unwrap_or("No further detail was provided.")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Boot diagnostic codes, paired with the wire value the frontend's union
    /// matches on.
    ///
    /// This table is the contract with `frontend/src/core/boot/diagnostics.ts`.
    /// A rename on either side is otherwise invisible until the diagnostics
    /// screen falls through to its "unrecognised code" branch — on the one
    /// screen that only renders when the user is already stuck.
    const ALL: [(DiagnosticCode, &str); 13] = [
        (DiagnosticCode::WorkspaceMissing, "workspace_missing"),
        (DiagnosticCode::WorkspaceUnwritable, "workspace_unwritable"),
        (DiagnosticCode::WorkspaceLocked, "workspace_locked"),
        (DiagnosticCode::SidecarSpawnFailed, "sidecar_spawn_failed"),
        (DiagnosticCode::SidecarExited, "sidecar_exited"),
        (DiagnosticCode::HandshakeTimeout, "handshake_timeout"),
        (DiagnosticCode::PortBindFailed, "port_bind_failed"),
        (DiagnosticCode::FfmpegMissing, "ffmpeg_missing"),
        (DiagnosticCode::FfmpegUnrunnable, "ffmpeg_unrunnable"),
        (DiagnosticCode::MigrationFailed, "migration_failed"),
        (
            DiagnosticCode::DatabaseNewerThanApp,
            "database_newer_than_app",
        ),
        (
            DiagnosticCode::WorkspaceRelocationFailed,
            "workspace_relocation_failed",
        ),
        (DiagnosticCode::Unknown, "unknown"),
    ];

    #[test]
    fn every_code_serialises_to_the_snake_case_the_frontend_expects() {
        for (code, expected) in ALL {
            assert_eq!(
                serde_json::to_string(&code).unwrap(),
                format!("\"{expected}\""),
                "{code:?} does not match the frontend union"
            );
        }
    }

    #[test]
    fn no_diagnostic_code_collides_with_an_api_error_code() {
        // Q124: two namespaces, no shared values, nothing is ever both. The
        // Python set is closed by Q67, so it can be written down here.
        const API_ERROR_CODES: [&str; 6] = [
            "validation_error",
            "not_found",
            "conflict",
            "workspace_error",
            "unauthorized",
            "internal_error",
        ];

        for (_, wire) in ALL {
            assert!(
                !API_ERROR_CODES.contains(&wire),
                "`{wire}` appears in both enums; they must stay disjoint"
            );
        }
    }

    #[test]
    fn blank_detail_is_dropped_rather_than_rendered_as_an_empty_block() {
        let diagnostic =
            Diagnostic::new(DiagnosticCode::SidecarExited, "It stopped.").with_detail("   \n ");
        assert_eq!(diagnostic.detail, None);
    }

    #[test]
    fn detail_is_kept_verbatim() {
        let stderr = "Traceback (most recent call last):\n  File \"x\"";
        let diagnostic =
            Diagnostic::new(DiagnosticCode::SidecarSpawnFailed, "…").with_detail(stderr);
        assert_eq!(diagnostic.detail.as_deref(), Some(stderr));
    }

    #[test]
    fn degraded_ffmpeg_codes_map_to_named_diagnostics() {
        let missing = diagnose_degraded_health(
            Some("ffmpeg_missing"),
            Some("Missing FFmpeg binary: C:/ffmpeg/ffmpeg.exe"),
        );
        assert_eq!(missing.code, DiagnosticCode::FfmpegMissing);

        let unrunnable = diagnose_degraded_health(
            Some("ffmpeg_unrunnable"),
            Some("ffmpeg.exe exited with code 1"),
        );
        assert_eq!(unrunnable.code, DiagnosticCode::FfmpegUnrunnable);
    }
}

File: frontend/src/core/boot/diagnostics.ts
/**
 * What each diagnostics code means to a user, and what they can do about it.
 *
 * Q124 fixes the nine original codes; ticket 29 adds `port_bind_failed`,
 * `ffmpeg_missing`, and `ffmpeg_unrunnable`. This is the presentation layer over them. The
 * action set is the reason `workspace_locked` was not merged into
 * `workspace_unwritable`: they need different buttons, and a merge would either
 * offer "Take over" on a permissions failure or hide it on a lock.
 *
 * `src-tauri/src/diagnostics.rs` has a test asserting the wire values match the
 * union in `bootState.ts`. This file is the other half — a code that arrives
 * without an entry here falls through to `UNRECOGNISED`, which is still a
 * usable screen rather than a blank one.
 */

import type { DiagnosticCode } from '@/core/boot/bootState';

export type DiagnosticAction =
  'retry' | 'choose-workspace' | 'forget-workspace' | 'take-over' | 'proceed-without-backup';

export interface DiagnosticPresentation {
  title: string;
  /** One paragraph. What happened, in the user's terms. */
  explanation: string;
  actions: DiagnosticAction[];
  /** Which log is likely to contain the cause. */
  log: 'boot' | 'backend';
}

export const UNRECOGNISED: DiagnosticPresentation = {
  title: 'Something stopped OutreachOS from starting',
  explanation:
    'The details below were written to the log. This is a failure the app does not have a specific explanation for.',
  actions: ['retry', 'choose-workspace'],
  log: 'boot',
};

const PRESENTATIONS: Record<DiagnosticCode, DiagnosticPresentation> = {
  workspace_missing: {
    title: 'The workspace folder is missing',
    explanation:
      'The folder OutreachOS was using is no longer where it was. It may have been renamed or moved, or it may be on a drive that is not connected.',
    // Q40: "Forget this workspace" is what stops this being a permanent boot
    // loop whose only exit is editing app-data by hand.
    actions: ['retry', 'choose-workspace', 'forget-workspace'],
    log: 'boot',
  },
  workspace_unwritable: {
    title: 'The workspace folder cannot be written to',
    explanation:
      'OutreachOS could not create a file in the workspace. This usually means the folder belongs to another user, or is inside a protected location such as Program Files.',
    actions: ['choose-workspace', 'retry'],
    log: 'boot',
  },
  workspace_locked: {
    title: 'The workspace is in use',
    explanation:
      'Another copy of OutreachOS appears to have this workspace open. If that copy is not running any more — for example, after a crash or on another machine — you can take it over.',
    actions: ['take-over', 'choose-workspace', 'retry'],
    log: 'boot',
  },
  sidecar_spawn_failed: {
    title: 'The backend could not be started',
    explanation:
      'OutreachOS could not launch its backend process at all. The log below records where it looked for it.',
    actions: ['retry'],
    log: 'boot',
  },
  sidecar_exited: {
    title: 'The backend stopped',
    explanation:
      'The backend process ended unexpectedly. Retrying starts a fresh one; nothing in the workspace has been changed.',
    actions: ['retry'],
    log: 'backend',
  },
  handshake_timeout: {
    title: 'The backend did not finish starting',
    explanation:
      'The backend was launched but did not become ready in time. The details below say how far it got.',
    actions: ['retry'],
    log: 'backend',
  },
  port_bind_failed: {
    title: 'The backend could not bind a local port',
    explanation:
      'OutreachOS could not open a loopback network port for its backend. Another program may be interfering, or the system may be out of available ports.',
    actions: ['retry'],
    log: 'backend',
  },
  ffmpeg_missing: {
    title: 'The bundled video tools are missing',
    explanation:
      'OutreachOS could not find FFmpeg or FFprobe where the installer placed them. Reinstalling the application may restore them.',
    actions: ['retry'],
    log: 'backend',
  },
  ffmpeg_unrunnable: {
    title: 'The bundled video tools could not be run',
    explanation:
      'FFmpeg or FFprobe is present but failed to start. The log below may show an antivirus block, a permissions problem, or corrupted binaries.',
    actions: ['retry'],
    log: 'backend',
  },
  migration_failed: {
    title: 'The database could not be updated',
    explanation:
      'OutreachOS could not apply a required database change. Your data has not been modified — the update was stopped before it began.',
    actions: ['retry', 'proceed-without-backup'],
    log: 'backend',
  },
  database_newer_than_app: {
    title: 'This workspace was made by a newer version',
    explanation:
      'The database in this workspace uses a newer format than this copy of OutreachOS understands. Opening it anyway could damage it, so it has not been opened. Install the newer version, or choose a different workspace.',
    // Deliberately no Retry: retrying does the same thing and fails the same
    // way. Offering it would suggest the problem might resolve itself.
    actions: ['choose-workspace', 'forget-workspace'],
    log: 'backend',
  },
  workspace_relocation_failed: {
    title: 'The workspace could not be moved',
    explanation:
      'OutreachOS stopped the backend and tried to switch workspace locations, but the move did not finish. Your original workspace is still the active one — nothing was pointed at the new folder.',
    actions: ['retry', 'choose-workspace'],
    log: 'boot',
  },
  unknown: UNRECOGNISED,
};

export function presentDiagnostic(code: DiagnosticCode): DiagnosticPresentation {
  return PRESENTATIONS[code] ?? UNRECOGNISED;
}
